### PR TITLE
Add torch.compiler.export_python: hill-climbable compiled kernels on disk

### PR DESCRIPTION
**Pull request by bobrenjc93**
`torch.compiler.export_python` wraps `torch.compiler.precompile` in a decorator keyed by a readable Python artifact on disk. The first call captures and lowers the decorated function, publishes the self-contained source, and runs it; every later call, in every later process, `exec`s that file.

```python
@torch.compiler.export_python(path="kernels/attn.py")
def attn(q, k, v):
    ...
```

## What this is for

**Hill climbing on generated kernels, and convenient kernel generation — not deployment.**

The emitted file is the source of truth and is always `exec`'d, so you can open it, retune a block size or a schedule, rerun, and your edit takes effect immediately. No recapture, no cache to defeat. The result is ordinary source that can be committed, diffed and reviewed.

That is the whole motivation, and the design follows from it rather than from a safety story.

## The central tradeoff: the artifact is sticky

`path` is the entire cache key. Nothing hashes `fn`'s body, the ambient config, or the machine.

That is **deliberate**: an artifact that re-captured whenever something changed would throw away your edits, which would defeat the point. The cost is real and worth stating plainly — **edit `fn`, rerun, and you get the old compiled code with no warning.** Delete the path to recapture.

Six comment stamps catch the ambient changes most likely to change the answer silently (module train/eval, input aliasing, duplicate input objects, autocast, and the globals the code generator resolves against). They are a **backstop, not a guarantee**: the set of things a generated kernel bakes is longer than the set that is checked.

So for the iteration loop there is an explicit verifier:

```bash
COMPILER_EXPORT_PYTHON_CHECK=1 python train.py
```

Every call then re-runs `fn` eagerly on a deep copy of the same inputs and compares, raising with the max absolute and relative difference if they diverge. Inputs are copied before the call because the graph may mutate them; the generators are rewound so a draw is not mistaken for an edit; and a `fn` that draws is skipped with a warning rather than reported as broken, since inductor's philox differs from eager's at the same seed by design. It doubles the per-call work — it is a debug mode, and it is the intended way to know an edit is safe.

## What was deliberately removed

**Buffer donation (`out=` / `unsafe_reduce_overhead`)** was implemented and then taken back out. It is a deployment-throughput feature with no bearing on hill climbing or kernel generation; it was the largest and most dangerous part of the implementation; and measurement showed it to be a net per-call *slowdown* for the shapes in the docstring's own example. Removing it deleted roughly 40% of the module and every memory-safety hazard that lived there.

Separately: inductor can emit the same `async_compile.triton` block twice under one name, and the later assignment wins. In a file whose purpose is being hand-edited that is the worst possible failure — you tune the first kernel you find and nothing happens — so byte-identical shadowed blocks are now dropped before publishing.

## Review order

1. `torch/compiler/__init__.py` — the public contract, motivation and tradeoff.
2. `torch/compiler/_export_python.py` — artifact lifecycle, the stamps, the verifier.
3. `torch/_precompile.py` — capture, RNG restore, the capture lock.
4. `test/test_precompile.py` — device-generic coverage.

## Known limitations

None of these are addressed here, and a reviewer should know them rather than discover them:

- A tensor subclass's inner shapes and a DTensor's placements are unrecorded, so a resharded input silently runs against capture's baked local shapes.
- `path` is not rank-aware, and a DTensor artifact hardcodes the c10d process-group name.
- A closed-over `torch.Generator` or torchbind object produces an artifact that can never bind it.
- An artifact is valid only on the machine type that produced it. Loading under a *wider* CPU ISA writes past the end of the output; under a *narrower* one it leaves part of the output uninitialized. Neither raises. `cpp.simdlen` and `ATEN_CPU_CAPABILITY` count as machine type.
- The per-call guard costs a multiple of the guarded work on small graphs.
- Stamp-missing warnings are not latched; a dangling symlink at `path` is misdiagnosed; a non-UTF-8 byte escapes as `UnicodeDecodeError`.

## Testing

```bash
python test/test_precompile.py
```

```bash
COMPILER_EXPORT_PYTHON_CHECK=1 python test/test_precompile.py
```

GLOO is present in this build, but an editable install registers no `torch.distributed.backends` entry point, so `init_process_group("gloo")` cannot resolve the name and two DTensor tests error locally. Registering it first makes the file pass — **321 tests, OK**:

```bash
python -c "import torch.distributed.distributed_c10d as c; c._register_builtin_gloo_backend(); import runpy, sys; sys.argv=['test/test_precompile.py']; runpy.run_path('test/test_precompile.py', run_name='__main__')"
```

Coverage was checked by mutation: reverting each fix to the code it replaced fails a test, verified against the actual prior revision rather than an invented weakening.

---

Developed with the assistance of an AI coding assistant.

### CODE REVIEW

**Comment by ezyang** [comment]
How does this relate to https://github.com/meta-pytorch/export-python ?

**Comment by bobrenjc93** [comment]
@ezyang this is still in the prototype stage, but the basic idea is:

1. it’s built on top of the existing precompile python codegen pipeline rather than relying on ast source transformations and manual rewriting of inductor output.
2. the ux is a bit different. it uses a sticky-cache style decorator that precompiles and generates the initial python file, after which agents can hill climb directly on that generated code. this also preserves the “pytorch as the reference language” model: if a researcher changes the computation, they can simply delete the generated file and regenerate a new kernel from the updated pytorch code.

**Comment by ezyang** [comment]
I think the UX change is a good direction.

Mmmm OK I guess it's not irrecoverably bad if this isn't in annex. Carry on

**Comment by bobrenjc93** [comment]
Superseded by the ghstack stack topped by #193901 (#193973, #193974, #193898, #193899, #193900, #193901), which carries this PR's export_python work re-sliced into reviewable commits plus the inductor readable-wrapper it depends on, rebased onto current main.
